src/drivers/src/USBDetection.cpp

```diff
@@ -0,0 +1,96 @@
+*/
+namespace USBDetection {
+
+    UsbDetector::UsbDetector() {
+        libusb_context *tmp;
+        libusb_init(&tmp);
+        ctx = ctx_t(tmp,libusb_exit);
+    }
+
+    std::set<usbdev_t> UsbDetector::find_arduinos() {
+        libusb_device **list;
+        cnt = libusb_get_device_list(ctx.get(), &list);
+
+        std::set<usbdev_t> arduinos;
+
+        int err = 0;
+        std::cout << cnt << std::endl;
+        if (cnt < 0) {
+            std::cerr << "error" << std::endl;
+        }
+        for (int i = 0; i < cnt; i++) {
+            struct libusb_device_descriptor info;
+            libusb_get_device_descriptor(list[i], &info);
+            if (info.idVendor == 0x2a03) { // 2a03 is arduino vendor id
+                libusb_ref_device(list[i]);
+                arduinos.insert(usbdev_t(list[i], libusb_unref_device));
+            }
+        }
+        libusb_free_device_list(list, 1);
+        return arduinos;
+    }
+
+    ArduinoConnection UsbDetector::open_arduino_with_prefix(std::string prefix) {
+        auto arduinos = find_arduinos();
+
+        size_t sz = arduinos.size();
+        std::vector<dhandle_t> handles(sz);
+        
+        int i = 0;
+        for (usbdev_t const &it : arduinos) {
+            libusb_device_handle *devhandle;
+            int err = libusb_open(it.get(), &devhandle);
+            if (err == 0) {
+                handles[i] = dhandle_t(devhandle, libusb_close);
+            } else {} //TODO
+            i++;
+        }
+
+        for (dhandle_t &it : handles) {
+            
+            libusb_device_handle *current_handle = it.get();
+
+            int numBytes;
+            unsigned char data[512];
+
+            int transfer = libusb_bulk_transfer(current_handle, LIBUSB_ENDPOINT_IN, data, sizeof(data), &numBytes, 0);
+
+            std::string dataStr((char *)data, 0, 10);
+            if (dataStr.compare(0, prefix.length(), prefix) == 0) {
+                return ArduinoConnection(std::move(it));
+            }
+        }
+        return ArduinoConnection();
+    }
+
+    ArduinoConnection UsbDetector::open_arduino_with_prefix(char *prefix) {
+        std::string prefixStr(prefix);
+        return open_arduino_with_prefix(prefixStr);
+    }
+
+
+    ArduinoConnection::ArduinoConnection(dhandle_t handle) {
+        arduino = std::move(handle); 
+    }
+
+    /*void ArduinoConnection::write(const std::string &msg) {
+        unsigned char *data = reinterpret_cast<unsigned char *>(const_cast<char *>((const_cast<std::string&>(msg)).c_str()));
+
+        ssize_t data_sz = msg.length();
+        int numBytes;
+        int transfer = libusb_bulk_transfer(arduino, LIBUSB_ENDPOINT_OUT, data, data_sz, &numBytes, 0);
+    }*/
+
+    std::string ArduinoConnection::read() {
+        unsigned char data[512];
+        int numBytes;
+        int transfer = libusb_bulk_transfer(arduino.get(), LIBUSB_ENDPOINT_IN, data, sizeof(data), &numBytes, 0);
+
+        return std::string(data, data+numBytes);
```

Comment:
@fhackett what function were you thinking of using? Looking at the [libusb doc](http://libusb.org/static/api-1.0/io.html) on synchronous and asynchronous transfers, the only synchronous transfer is libusb_bulk_transfer. 

While I agree that using an asynchronous transfer is probably the best for our purposes, it states in the doc that "it is a more complex interface" (and looking at some examples seems to support this). I'm mostly trying to avoid any asynchrony not internal to the class, as it will likely be used in the future by **extremely** inexperienced devs, who may not be familiar with it.

I only have a very rudimentary understanding of this, and I'm likely misunderstanding some things, so if you have a second, could you perhaps do a branch of this branch and write a rough implementation of what you're thinking of? No pressure to do so, it would just help me understand a bit. 

Otherwise, going forward, there are a few options, and seeing as Em's responsible for this, it's totally up to him what he wants to do. (That being said, the Jan 15th deadline for a rough working product is still in place)

a) re-write this to use libusb's asynchronous interface
b) use bulk_transfer for now, then make (a) an issue to be addressed in another PR